src/Raven.Server/ServerWide/ClusterStateMachine.License.cs

```diff
@@ -0,0 +1,207 @@
+                    return;
+
+                throw new LicenseLimitException(LimitType.TimeSeriesRollupsAndRetention, "Your license doesn't support adding Time Series Rollups And Retention feature.");
+
+            case nameof(EditDocumentsCompressionCommand):
+                if (CanAssertLicenseLimits(context, minBuildVersion: MinBuildVersion54116) == false)
+                    return;
+
+                if (serverStore.LicenseManager.LicenseStatus.HasDocumentsCompression)
+                    return;
+
+                if (databaseRecord.DocumentsCompression.CompressAllCollections == false && databaseRecord.DocumentsCompression.CompressRevisions == false)
+                    return;
+
+                throw new LicenseLimitException(LimitType.DocumentsCompression, "Your license doesn't support adding Documents Compression feature.");
+
+            case nameof(AddElasticSearchEtlCommand):
+                if (CanAssertLicenseLimits(context, minBuildVersion: MinBuildVersion54116) == false)
+                    return;
+
+                if (serverStore.LicenseManager.LicenseStatus.HasElasticSearchEtl)
+                    return;
+
+                throw new LicenseLimitException(LimitType.ElasticSearchEtl, "Your license doesn't support adding Elastic Search ETL feature.");
+
+            case nameof(AddOlapEtlCommand):
+                if (CanAssertLicenseLimits(context, minBuildVersion: MinBuildVersion54116) == false)
+                    return;
+
+                if (serverStore.LicenseManager.LicenseStatus.HasOlapEtl)
+                    return;
+
+                throw new LicenseLimitException(LimitType.OlapEtl, "Your license doesn't support adding Olap ETL feature.");
+
+            case nameof(AddQueueEtlCommand):
+                if (CanAssertLicenseLimits(context, minBuildVersion: MinBuildVersion54116) == false)
+                    return;
+
+                if (serverStore.LicenseManager.LicenseStatus.HasQueueEtl)
+                    return;
+
+                throw new LicenseLimitException(LimitType.QueueEtl, "Your license doesn't support adding Queue ETL feature.");
+        }
+    }
+
+
+    private bool CanAssertLicenseLimits(ClusterOperationContext context, int minBuildVersion)
+    {
+        var licenseLimitsBlittable = Read(context, ServerStore.LicenseLimitsStorageKey);
+        if (licenseLimitsBlittable == null)
+            return false;
+
+        var licenseLimits = JsonDeserializationServer.LicenseLimits(licenseLimitsBlittable);
+        if (licenseLimits.NodeLicenseDetails == null)
+            return false;
+
+        var clusterTopology = _parent.GetTopology(context);
+
+        foreach (var clusterNode in clusterTopology.Members.Union(clusterTopology.Watchers).Union(clusterTopology.Promotables))
+        {
+            if (licenseLimits.NodeLicenseDetails.ContainsKey(clusterNode.Key) == false)
```

Comment:
What does this check mean?

Reply:
This means that the node details aren't available right now (a new node for example).

Reply:
so if you are adding a new node (even a non-existing one) you skip any license check? why?

Reply:
Wonder of you should do this check only on cluster `Members` (voters) nodes

Reply:
Done